modules/src/cep18_token.rs

```diff
@@ -0,0 +1,399 @@
+//! CEP-18 Casper Fungible Token standard implementation.
+use odra::prelude::*;
+use odra::{casper_types::U256, Address, Mapping, SubModule, UnwrapOrRevert, Var};
+
+use crate::cep18::errors::Error;
+use crate::cep18::errors::Error::{
+    CannotTargetSelfUser, InsufficientRights, InvalidBurnTarget, MintBurnDisabled, Overflow
+};
+use crate::cep18::events::{
+    Burn, ChangeSecurity, DecreaseAllowance, IncreaseAllowance, Mint, SetAllowance, Transfer,
+    TransferFrom
+};
+use crate::cep18::storage::{
+    Cep18AllowancesStorage, Cep18BalancesStorage, Cep18DecimalsStorage, Cep18NameStorage,
+    Cep18SymbolStorage, Cep18TotalSupplyStorage
+};
+use crate::cep18::utils::{Cep18Modality, SecurityBadge};
+
+/// CEP-18 token module
+#[odra::module(events = [Mint, Burn, SetAllowance, IncreaseAllowance, DecreaseAllowance, Transfer,
+    TransferFrom, ChangeSecurity])]
+pub struct Cep18 {
+    decimals: SubModule<Cep18DecimalsStorage>,
+    symbol: SubModule<Cep18SymbolStorage>,
+    name: SubModule<Cep18NameStorage>,
+    total_supply: SubModule<Cep18TotalSupplyStorage>,
+    balances: SubModule<Cep18BalancesStorage>,
+    allowances: SubModule<Cep18AllowancesStorage>,
+    security_badges: Mapping<Address, SecurityBadge>,
+    modality: Var<u8>
+}
+
+#[odra::module]
+impl Cep18 {
+    /// Initializes the contract with the given metadata, initial supply, security and modality.
+    #[allow(clippy::too_many_arguments)]
+    pub fn init(
+        &mut self,
+        symbol: String,
+        name: String,
+        decimals: u8,
+        initial_supply: U256,
+        minter_list: Vec<Address>,
+        admin_list: Vec<Address>,
+        modality: Option<u8>
```

Comment:
As `Cep18Modality` is odra type, and its `CLType` is `U8` I'd use something more meaningful than raw u8.
```suggestion
        modality: Option<Cep18Modality>
```